### PR TITLE
script to Simulate DVM core logic post UMIP-3 upgrade

### PR DESCRIPTION
**Pull request by chrismaree**
This script introduces some additional simulation logic to the UMIP-3 upgrade to test the deployment of an expiring multi-party contract. 

The logic was not changed from the UMPI-3 upgrade 2 weeks ago. This PR pushes up the code so we have it for testing the next DVM upgrade.

### CODE REVIEW

**Comment by chrismaree** [comment]
@nicholaspai @ptare @mrice32 Please take a look so we can get this merged. Nothing has changed really from when this script was created before the upgrade other than a clean.

File: core/scripts/umip-3/2_VoteSimulate.js [hunk 1]
```diff
@@ -36,9 +44,9 @@ async function runExport() {
   if (argv.repeated) {
     assert((await governor.numProposals()).toNumber() > 1);
   } else {
-    assert.equal((await governor.numProposals()).toNumber(), 1); // there should be 1 proposal on the first run.
+    assert.equal((await governor.numProposals()).toNumber(), 2); // there should be 1 proposal on the first run.
```
Comment (nicholaspai) [inline core/scripts/umip-3/2_VoteSimulate.js:47]: Should there still be 1 proposal at the start?

File: core/scripts/umip-3/2_VoteSimulate.js [hunk 2]
```diff
@@ -36,9 +44,9 @@ async function runExport() {
...
   }
-  assert.equal((await voting.getPendingRequests()).length, 0); // There should be no pending requests
+  assert.equal((await voting.getPendingRequests()).length, 1); // There should be no pending requests
```
Comment (nicholaspai) [inline core/scripts/umip-3/2_VoteSimulate.js:49]: Same comment

File: core/scripts/umip-3/2_VoteSimulate.js [hunk 3]
```diff
@@ -72,7 +80,7 @@ async function runExport() {
     (await voting.getCurrentRoundId()).toString()
   );
   let pendingRequests = await voting.getPendingRequests();
-  assert.equal(pendingRequests.length, 1); // the one proposal should have advanced to a request
+  assert.equal(pendingRequests.length, 2); // the one proposal should have advanced to a request
```
Comment (nicholaspai) [inline core/scripts/umip-3/2_VoteSimulate.js:83]: Same comment

File: core/scripts/umip-3/2_VoteSimulate.js [hunk 4]
```diff
@@ -153,14 +161,14 @@ async function runExport() {
     (await voting.getCurrentRoundId()).toString()
   );
 
-  assert.equal((await voting.getPendingRequests()).length, 0); // There should be no pending requests as vote is concluded
+  assert.equal((await voting.getPendingRequests()).length, 1); // There should be no pending requests as vote is concluded
```
Comment (nicholaspai) [inline core/scripts/umip-3/2_VoteSimulate.js:164]: Same comment

File: core/scripts/umip-3/2_VoteSimulate.js [hunk 5]
```diff
@@ -153,14 +161,14 @@ async function runExport() {
...
 
   /** *******************************************************************
    * 4) Execute proposal submitted to governor now that voting is done *
    **********************************************************************/
 
   console.log("4. EXECUTING GOVERNOR PROPOSALS");
-  const proposalId = (await governor.numProposals()).subn(1).toString(); // most recent proposal in voting.sol
+  const proposalId = (await governor.numProposals()).subn(2).toString(); // most recent proposal in voting.sol
```
Comment (nicholaspai) [inline core/scripts/umip-3/2_VoteSimulate.js:171]: Same comment

File: core/scripts/umip-3/3_Verify.js [hunk 6]
```diff
@@ -35,12 +41,12 @@ async function runExport() {
   console.log(" 1. Validating deployed bytecode at new addresses...");
 
   // The deployed bytecode should match the expected bytecode for all new contracts deployed.
-  await compiledByteCodeMatchesDeployed(Voting);
-  await compiledByteCodeMatchesDeployed(Registry);
-  await compiledByteCodeMatchesDeployed(Store);
-  await compiledByteCodeMatchesDeployed(FinancialContractsAdmin);
-  await compiledByteCodeMatchesDeployed(IdentifierWhitelist);
-  await compiledByteCodeMatchesDeployed(Governor);
+  // await compiledByteCodeMatchesDeployed(Voting);
```
Comment (nicholaspai) [inline core/scripts/umip-3/3_Verify.js:44]: Should we just get rid of this script now if we're not verifying bytecode?

File: core/scripts/umip-3/4_ExecuteSimulate.js [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+  );
+
+  assert.equal((await voting.getPendingRequests()).length, 0); // There should be no pending requests as vote is concluded
+
+  /** ******************************************************************
+   * 8) Execute proposal submitted to governor now that voting is done *
+   *********************************************************************/
+  console.log("8. EXECUTING GOVERNOR PROPOSALS");
+  console.log((await governor.numProposals()).toNumber());
+  const proposalId = 0; // first and only proposal in voting.sol
```
Comment (nicholaspai) [inline core/scripts/umip-3/4_ExecuteSimulate.js:248]: Just confirm that this is the correct proposal index.

File: core/scripts/umip-3/4_ExecuteSimulate.js [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+    "currentRoundId",
+    (await voting.getCurrentRoundId()).toString()
+  );
+
+  assert.equal((await voting.getPendingRequests()).length, 0); // There should be no pending requests as vote is concluded
+
+  /** ****************************
+   * 14) Retrieving rewards from voting *
+   *********************************/
+  console.log("14. RETRIEVING DVM VOTING REWARDS");
```
Comment (nicholaspai) [inline core/scripts/umip-3/4_ExecuteSimulate.js:462]: You should remove this section or write a TODO

File: core/scripts/umip-3/2_VoteSimulate.js (hunk 1 above)
Comment (chrismaree) [reply core/scripts/umip-3/2_VoteSimulate.js:47]: there should be 2. have updated the comment accordingly.

File: core/scripts/umip-3/2_VoteSimulate.js (hunk 2 above)
Comment (chrismaree) [reply core/scripts/umip-3/2_VoteSimulate.js:49]: there should be no pending proposals. have updated accordingly

File: core/scripts/umip-3/3_Verify.js (hunk 6 above)
Comment (chrismaree) [reply core/scripts/umip-3/3_Verify.js:44]: we can. I thought we could leave it for now as we might at some point want to verify this in some way.

File: core/scripts/umip-3/2_VoteSimulate.js (hunk 5 above)
Comment (nicholaspai) [reply core/scripts/umip-3/2_VoteSimulate.js:171]: @chrismaree just in case you missed this

File: core/scripts/umip-3/2_VoteSimulate.js (hunk 4 above)
Comment (nicholaspai) [reply core/scripts/umip-3/2_VoteSimulate.js:164]: @chrismaree just in case you missed this

File: core/scripts/umip-3/2_VoteSimulate.js (hunk 3 above)
Comment (nicholaspai) [reply core/scripts/umip-3/2_VoteSimulate.js]: @chrismaree just in case you missed this

File: core/scripts/umip-3/2_VoteSimulate.js (hunk 5 above)
Comment (chrismaree) [reply core/scripts/umip-3/2_VoteSimulate.js:171]: ye it's correct now I think :) the problem is that this script will need to be updated according to the specific vote we are acting on and as such the as the DVM's state due to proposals/votes etc it'll need to be adapted accordingly.

Comment (nicholaspai) [reply core/scripts/umip-3/2_VoteSimulate.js:171]: Yeah which makes this less useful for the future. Is there any way to read it dynamically

Comment (nicholaspai) [reply core/scripts/umip-3/2_VoteSimulate.js:171]: You can at least make this one less than the number hardcoded into line 47.

```
const startingProposals = 2
assert.equal(...numProposals.length, startingProposals+1)
const proposalId = (...numProposals()).subn(startingProposals).toString()
```

File: core/scripts/umip-3/4_ExecuteSimulate.js (hunk 7 above)
Comment (chrismaree) [reply core/scripts/umip-3/4_ExecuteSimulate.js:248]: yup. it is as this script assumes that the upgrade has deployed a new DVM.

**Comment by chrismaree** [comment]
@nicholaspai this PR is ready to go. the specific parameters in the scripts don't need to be updated past where they are now as they'll need to be tuned in any case when we use this script again in the future. Let's get this merged.

**Pull request merged**